source/common/router/router_ratelimit.cc

```diff
@@ -107,30 +108,27 @@ void RateLimitPolicyEntryImpl::populateDescriptors(
   }
 }
 
-RateLimitPolicyImpl::RateLimitPolicyImpl(const Json::Object& config) {
+RateLimitPolicyImpl::RateLimitPolicyImpl(const Json::Object& config)
+    : rate_limit_entries_reference_(RateLimitPolicyImpl::MAX_STAGE_NUMBER + 1) {
   if (config.hasObject("rate_limits")) {
-    std::vector<std::unique_ptr<RateLimitPolicyEntry>> rate_limit_policy;
-    std::vector<std::reference_wrapper<const RateLimitPolicyEntry>> rate_limit_policy_reference;
     for (const Json::ObjectPtr& rate_limit : config.getObjectArray("rate_limits")) {
       std::unique_ptr<RateLimitPolicyEntry> rate_limit_policy_entry(
           new RateLimitPolicyEntryImpl(*rate_limit));
-      rate_limit_policy_reference.emplace_back(*rate_limit_policy_entry);
-      rate_limit_policy.emplace_back(std::move(rate_limit_policy_entry));
+      uint64_t stage = rate_limit_policy_entry->stage();
+      ASSERT(stage < rate_limit_entries_reference_.size());
+      rate_limit_entries_reference_[stage].emplace_back(*rate_limit_policy_entry);
+      rate_limit_entries_.emplace_back(std::move(rate_limit_policy_entry));
     }
-    rate_limit_entries_.emplace_back(std::move(rate_limit_policy));
-    rate_limit_entries_reference_.emplace_back(rate_limit_policy_reference);
   }
 }
 
-const std::vector<std::reference_wrapper<const RateLimitPolicyEntry>>&
-    RateLimitPolicyImpl::getApplicableRateLimit(int64_t) const {
-  // Currently return all rate limit policy entries.
-  // TODO(mattklein123): Implement returning only rate limit policy entries that match the stage
-  // setting.
+const std::vector<std::reference_wrapper<const Router::RateLimitPolicyEntry>>&
+RateLimitPolicyImpl::getApplicableRateLimit(uint64_t stage) const {
+  ASSERT(stage <= RateLimitPolicyImpl::MAX_STAGE_NUMBER);
   if (rate_limit_entries_.empty()) {
```

Comment:
Do we even need empty_rate_limit_ anymore? Can't we just delete it and always return the stage vector which should be correctly empty?